sensor_msgs_py/sensor_msgs_py/point_cloud2.py

```diff
@@ -68,80 +81,72 @@ def read_points(cloud, field_names=None, skip_nans=False, uvs=[]):
-    else:
-        if uvs:
-            for u, v in uvs:
-                yield unpack_from(data, (row_step * v) + (point_step * u))
-        else:
-            for v in range(height):
-                offset = row_step * v
-                for u in range(width):
-                    yield unpack_from(data, offset)
-                    offset += point_step
+    return iter(read_points_list(cloud, field_names=field_names, skip_nans=skip_nans, uvs=[]))
 
 
 def read_points_list(cloud, field_names=None, skip_nans=False, uvs=[]):
     """
     Read points from a sensor_msgs.PointCloud2 message.
 
-    This function returns a list of namedtuples. It operates on top of the
-    read_points method. For more efficient access use read_points directly.
-
-    :param cloud: The point cloud to read from. (Type: sensor_msgs.PointCloud2)
+    :param cloud: The point cloud to read from sensor_msgs.PointCloud2.
     :param field_names: The names of fields to read. If None, read all fields.
                         (Type: Iterable, Default: None)
     :param skip_nans: If True, then don't return any point with a NaN value.
                       (Type: Bool, Default: False)
     :param uvs: If specified, then only return the points at the given
-                coordinates. (Type: Iterable, Default: empty list]
-    :return: List of namedtuples containing the values for each point
+        coordinates. (Type: Iterable, Default: empty list)
+    :return: numpy.recarray with values for each point.
     """
     assert isinstance(cloud, PointCloud2), \
         'cloud is not a sensor_msgs.msg.PointCloud2'
 
-    if field_names is None:
-        field_names = [f.name for f in cloud.fields]
+    assert cloud.point_step * cloud.width * cloud.height == len(cloud.data), \
+        'length of data does not match point_step * width * height'
+
+    all_field_names = []
+    np_struct = []
+    total_used_bytes = 0
+    for field in cloud.fields:
+        all_field_names.append(field.name)
+        assert field.datatype in _DATATYPES_NUMPY_MAP, \
+            'invalid datatype %d specified for field %s' % (field.datatype, field.name)
+        field_np_datatype = _DATATYPES_NUMPY_MAP[field.datatype]
+        np_struct.append((field.name, field_np_datatype))
+        total_used_bytes += np.nbytes[field_np_datatype]
+
+    assert cloud.point_step >= total_used_bytes, \
+        'error: total byte sizes of fields exceeds point_step'
+
+    if cloud.point_step > total_used_bytes:
+        np_struct.append(('unused_bytes', np.uint8, cloud.point_step - total_used_bytes))
+
+    points = np.frombuffer(cloud.data, dtype=np_struct).view(dtype=np.recarray)
 
-    Point = namedtuple('Point', field_names)
+    if skip_nans:
+        nan_indexes = None
+        for field_name in all_field_names:
+            if nan_indexes is None:
+                nan_indexes = np.isnan(points[field_name])
+            else:
+                nan_indexes = nan_indexes | np.isnan(points[field_name])
```

Comment:
@dheera nit^2: 
```suggestion
        nan_indexes = False
        for field_name in all_field_names:
            nan_indexes |= np.isnan(points[field_name])
```

Reply:
Also, if `field_names` is not `None`, we can probably filter columns out before doing this.